Config redo

benblack769 opened the pull request.

benblack769 committed:
docs started, old tests deleted

benblack769 committed:
comments added and private.py reorganized

benblack769 committed:
simplified and unified configuration

benblack769 committed:
Merge branch 'master' into config_redo

kamba16 commented:
1. so you moved `config_canyon.py` into `config.py`
2. you also made a dictionary object `straboconfig` that is populated in runserver.py
   Is there a way for us to clarify that it is a dictionary in the name? like config_dict or something. Idk that confused me the first time I saw it. 
3. why did you change `**app.config` to `straboconfig=straboconfig` in the views files?
4. And you consolidated parts of `image_processing.py` into `file_writing.py` (??)

Overall, what's the benefit of creating the dictionary `straboconfig` versus using the `app.config`? Is that best practice? I ask because you set it to access the dictionary in the HTML files in between double brackets {{}} which I haven't seen done before. the original code is more common online, I think.

benblack769 committed:
Merge branch 'master' into config_redo

benblack769 committed:
templates changed back

benblack769 commented:
Answering questions, 
1. Yeah.
2. Doing something like that is generally considered "unpythonic". I sort of like it, but it makes the idea of dynamic typing kind of pointless. One real problem with doing this is that objects may not quite be dictionaries, but are used like them anyways. For example, `request.form` is not a dictionary at all, but a special construct called a multidict, which flask gets from the werkzeug library, which you can use just like a regular dictionary and it works fine, but you can also do many other things with. In fact `app.config` is not quite a dictionary either. The point being, in python, it really doesn't matter what type things are, but what you can do with them. And in python, you typically can do that just by looking at the code. In the python world, you can generally assume (and people will assume) that anything that looks like `cont[obj] = other_obj` where obj is not an integer, serves the same purpose as a dictionary. Of course, there are exceptions to this general ideology, especially when objects get really complicated and weird, but with dictionaries, this idea holds.
3. Changed this.
4. Can you explain further what you are confused about here? Also, I did not change those files substantially in this pull request, I changed those before you left.

The biggest advantage to keeping them separate is the ability to send over entire straboconfig dictionary over to the javascript in one go. Before, it was impossible to do this because app.config comes will all sorts of default values that cannot be jsonified.

benblack769 committed:
merged with bencomments

benblack769 committed:
fixed merge problems

kamba16 commented:
There are a couple of files (`footer.html`,`header.html`) where you did not change the templates back to the double brackets {{}}. That's okay, though, once we merge the pull request I can go through and double check what we have. 

I wish our `straboconfig` dictionary was more clearly a dictionary, but if our documentation is **really** clear then that should not be a problem. 

I am comfortable merging these changes.
